apm-agent-core/src/main/java/co/elastic/apm/agent/impl/CloudMetadataProvider.java

```diff
@@ -0,0 +1,401 @@
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+import static co.elastic.apm.agent.configuration.CoreConfiguration.CloudProvider.AWS;
+import static co.elastic.apm.agent.configuration.CoreConfiguration.CloudProvider.AZURE;
+import static co.elastic.apm.agent.configuration.CoreConfiguration.CloudProvider.GCP;
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+public class CloudMetadataProvider {
+    private static final Logger logger = LoggerFactory.getLogger(CloudMetadataProvider.class);
+
+    private static final DslJson<Object> dslJson = new DslJson<>(new DslJson.Settings<>());
+
+    /**
+     * Automatic discovery of cloud provider and related metadata. This method is fetching data from public APIs
+     * exposed by known cloud provider services through HTTP.
+     * When a specific {@link co.elastic.apm.agent.configuration.CoreConfiguration.CloudProvider} is specified, only
+     * the relevant API will be queried.
+     * However, when called with {@link co.elastic.apm.agent.configuration.CoreConfiguration.CloudProvider#AUTO}, all
+     * known endpoints are queried concurrently. In such cases, blocking is expected to be long, bounded by the
+     * HTTP requests timing out.
+     *
+     * @param cloudProvider  the expected {@link co.elastic.apm.agent.configuration.CoreConfiguration.CloudProvider}
+     * @param queryTimeoutMs timeout in milliseconds to limit the discovery duration
+     * @return Automatically discovered {@link CloudProviderInfo}, or {@code null} if none found.
+     */
+    @Nullable
+    static CloudProviderInfo fetchAndParseCloudProviderInfo(final CoreConfiguration.CloudProvider cloudProvider, final int queryTimeoutMs) {
+
+        Throwable unexpectedError = null;
+        CloudProviderInfo cloudProviderInfo = null;
+        try {
+            switch (cloudProvider) {
+                case AWS: {
+                    try {
+                        cloudProviderInfo = getAwsMetadata(queryTimeoutMs, cloudProvider);
+                    } catch (Exception e) {
+                        unexpectedError = e;
+                    }
+                    break;
+                }
+                case GCP: {
+                    try {
+                        cloudProviderInfo = getGcpMetadata(queryTimeoutMs);
+                    } catch (Exception e) {
+                        unexpectedError = e;
+                    }
+                    break;
+                }
+                case AZURE: {
+                    try {
+                        cloudProviderInfo = getAzureMetadata(queryTimeoutMs);
+                    } catch (Exception e) {
+                        unexpectedError = e;
+                    }
+                    break;
+                }
+                case AUTO: {
```

Comment:
[minor] maybe move this large block into a separate method as it's already nested in a switch statement.